From b077437f2a093bdd8758da6859f6df1253b5985a Mon Sep 17 00:00:00 2001
From: Noah Anleitner <noah.anleitner@nonetwor.cc>
Date: Sun, 11 Oct 2026 19:33:16 +0200
Subject: [PATCH] chibios: add the LPC546xx family, for the LPC54605

Everything the family needs is discovered from these variables, so no other
build file changes: platform.mk already searches ChibiOS-Contrib for both the
startup makefile and os/hal/ports/$(MCU_FAMILY)/$(MCU_SERIES)/platform.mk.

The HAL port itself lives in ChibiOS-Contrib. It also needs the family seam
patch to the shared LPC USB driver, which is not in Contrib yet.
---
 platforms/chibios/mcu_selection.mk | 43 ++++++++++++++++++++++++++++++
 1 file changed, 43 insertions(+)

diff --git a/platforms/chibios/mcu_selection.mk b/platforms/chibios/mcu_selection.mk
index 199bdb2..3a77e58 100644
--- a/platforms/chibios/mcu_selection.mk
+++ b/platforms/chibios/mcu_selection.mk
@@ -176,6 +176,49 @@ ifneq ($(findstring RP2040, $(MCU)),)
   FIRMWARE_FORMAT ?= uf2
 endif
 
+ifneq ($(findstring LPC54605, $(MCU)),)
+  # Cortex version
+  MCU = cortex-m4
+
+  # ARM version, CORTEX-M0/M1 are 6, CORTEX-M3/M4/M7 are 7
+  ARMV = 7
+
+  ## chip/board settings
+  # - the next two should match the directories in
+  #   <chibios[-contrib]>/os/hal/ports/$(MCU_PORT_NAME)/$(MCU_SERIES)
+  #   OR
+  #   <chibios[-contrib]>/os/hal/ports/$(MCU_FAMILY)/$(MCU_SERIES)
+  MCU_FAMILY = LPC
+  MCU_SERIES = LPC546xx
+
+  # Linker script to use
+  # - it should exist either in <chibios>/os/common/startup/ARMCMx/compilers/GCC/ld/
+  #   or <keyboard_dir>/ld/
+  MCU_LDSCRIPT ?= LPC54605J512
+
+  # Startup code to use
+  #  - it should exist in <chibios>/os/common/startup/ARMCMx/compilers/GCC/mk/
+  MCU_STARTUP ?= lpc546xx
+
+  # Board: it should exist either in <chibios>/os/hal/boards/,
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= GENERIC_LPC_LPC54605
+
+  # Cortex-M4F, and the port's clock tree assumes the FPU is usable.
+  USE_FPU ?= yes
+
+  # NXP's headers want CPU_<part><package>; ChibiOS-Contrib's lpc_registry.h
+  # wants a bare LPC<part> that NXP's headers never define. Both have to be on
+  # the command line, because cmparams.h is reached from crt1.c and from
+  # vectors_lpc.S, where no board header is in scope.
+  LPC_PART ?= LPC54605
+  OPT_DEFS += -DCPU_LPC54605J512ET100 -D$(LPC_PART)
+
+  # Flashing is a host-side operation over Corsair's Bragi protocol; there is no
+  # self-hosted bootloader to jump to.
+  BOOTLOADER ?= custom
+endif
+
 ifneq ($(findstring STM32F042, $(MCU)),)
   # Cortex version
   MCU = cortex-m0
-- 
2.55.0

